### PR TITLE
[mlir][sparse][pybind][CAPI] remove LevelType enum from CAPI, constru…

### PR DESCRIPTION
**Pull request by PeimingLiu**
…ct LevelType from LevelFormat and properties instead.

**Rationale**
We used to explicitly declare every possible combination between `LevelFormat` and `LevelProperties`, and it now becomes difficult to scale as more properties/level formats are going to be introduced.

### CODE REVIEW

File: mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorAttrDefs.td [hunk 1]
```diff
@@ -162,6 +162,8 @@ def SparseTensorEncodingAttr : SparseTensor_Attr<"SparseTensorEncoding",
 
     - **nonunique** : duplicate coordinates may appear at the level
     - **nonordered** : coordinates may appear in arbribratry order
+    - **soa** : coordinates are fused into the parent level in SoA (structure of array)
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/SparseTensorAttrDefs.td:165]: You will have to add some text around L158 that by default, singleton is stored in AoS format, but that this default can be overwritten with the SoA property (similar to how we say that by default, stuff is ordered / unique and we can override that)

**Review by aartbik — COMMENTED** [review]
please add SoA/AoS in description, so we now when this was introduced

File: mlir/include/mlir/Dialect/SparseTensor/IR/Enums.h [hunk 2]
```diff
@@ -31,6 +31,7 @@
 // NOTE: Client code will need to include "mlir/ExecutionEngine/Float16bits.h"
 // if they want to use the `MLIR_SPARSETENSOR_FOREVERY_V` macro.
 
+#include "llvm/ADT/ArrayRef.h"
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/Enums.h]: We can't include anything llvm/mlir related here (see comment at L21)

File: mlir/include/mlir/Dialect/SparseTensor/IR/Enums.h [hunk 3]
```diff
@@ -343,17 +344,31 @@ constexpr std::optional<LevelFormat> getLevelFormat(LevelType lt) {
 /// Convert a LevelFormat to its corresponding LevelType with the given
 /// properties. Returns std::nullopt when the properties are not applicable
 /// for the input level format.
-constexpr std::optional<LevelType> buildLevelType(LevelFormat lf, bool ordered,
-                                                  bool unique, uint64_t n = 0,
-                                                  uint64_t m = 0) {
+inline std::optional<LevelType>
+buildLevelType(LevelFormat lf,
+               llvm::ArrayRef<LevelPropertyNondefault> properties,
```
Comment (aartbik) [inline mlir/include/mlir/Dialect/SparseTensor/IR/Enums.h]: can we use something different from array ref?

File: mlir/lib/Bindings/Python/DialectSparseTensor.cpp [hunk 4]
```diff
@@ -113,16 +109,14 @@ static void populateDialectSparseTensorSubmodule(const py::module &m) {
...
           })
-      .def_property_readonly("lvl_types_enum", [](MlirAttribute self) {
+      .def_property_readonly("lvl_formats_enum", [](MlirAttribute self) {
         const int lvlRank = mlirSparseTensorEncodingGetLvlRank(self);
-        std::vector<MlirBaseSparseTensorLevelType> ret;
+        std::vector<MlirSparseTensorLevelFormat> ret;
         ret.reserve(lvlRank);
         for (int l = 0; l < lvlRank; l++) {
           // Convert level type to 32 bits to ignore n and m for n_out_of_m
           // format.
```
Comment (yinying-lisa-li) [inline mlir/lib/Bindings/Python/DialectSparseTensor.cpp:118]: Can you remove this comment?

File: mlir/include/mlir-c/Dialect/SparseTensor.h [hunk 5]
```diff
@@ -27,23 +27,19 @@ MLIR_DECLARE_CAPI_DIALECT_REGISTRATION(SparseTensor, sparse_tensor);
...
-  MLIR_SPARSE_TENSOR_LEVEL_SINGLETON_NO = 0x000000040002,
-  MLIR_SPARSE_TENSOR_LEVEL_SINGLETON_NU_NO = 0x000000040003,
   MLIR_SPARSE_TENSOR_LEVEL_LOOSE_COMPRESSED = 0x000000080000,
-  MLIR_SPARSE_TENSOR_LEVEL_LOOSE_COMPRESSED_NU = 0x000000080001,
-  MLIR_SPARSE_TENSOR_LEVEL_LOOSE_COMPRESSED_NO = 0x000000080002,
-  MLIR_SPARSE_TENSOR_LEVEL_LOOSE_COMPRESSED_NU_NO = 0x000000080003,
   MLIR_SPARSE_TENSOR_LEVEL_N_OUT_OF_M = 0x000000100000,
 };
 
+enum MlirSparseTensorLevelProperty {
```
Comment (aartbik) [inline mlir/include/mlir-c/Dialect/SparseTensor.h:38]: LevelPropertyNondefault to be consistent with Enums

**Review by aartbik — APPROVED** [review]

**Review by yinying-lisa-li — APPROVED** [review]

**Pull request merged**